crypto/pem/pem_passwd.c

```diff
@@ -0,0 +1,301 @@
+# else /* _WIN32 */
+#  include <windows.h>
+# endif
+
+# define NUM_SIG 32
+
+static volatile sig_atomic_t intr_signal;
+static struct CRYPTO_STATIC_MUTEX console_global_mutex = CRYPTO_STATIC_MUTEX_INIT;
+
+# ifdef SIGACTION
+static struct sigaction savsig[NUM_SIG];
+# else
+static void (*savsig[NUM_SIG]) (int);
+# endif
+
+static int read_till_nl(FILE * in);
+static void recsig(int signal);
+static void pushsig(void);
+static void popsig(void);
+
+#if defined(_WIN32)
+  DWORD tty_orig, tty_new;
+#else
+  TTY_STRUCT tty_orig, tty_new;
+#endif
+
+FILE *tty_in, *tty_out;
+int is_a_tty;
+
+
+/* Internal functions to read a string without echoing */
+static int read_till_nl(FILE *in) {
+#  define SIZE 4
+    char buf[SIZE + 1];
+
+    do {
+        if (!fgets(buf, SIZE, in)) {
+            return 0;
+        }
+    } while (strchr(buf, '\n') == NULL);
+    return 1;
+}
+
+/* Signal handling functions */
+static void pushsig(void) {
+# if !defined(_WIN32)
+    struct sigaction sa;
+    OPENSSL_cleanse(&sa, sizeof(sa));
+
+    sa.sa_handler = recsig;
+
+    for (int i = 1; i < NUM_SIG; i++) {
+        if (i == SIGUSR1 || i == SIGUSR2 || i == SIGKILL) {
+            continue;
+        }
+        sigaction(i, &sa, &savsig[i]);
+    }
+# else // Specific error codes for windows
+    savsig[SIGABRT] = signal(SIGABRT, recsig);
+    savsig[SIGFPE] = signal(SIGFPE, recsig);
+    savsig[SIGILL] = signal(SIGILL, recsig);
+    savsig[SIGINT] = signal(SIGINT, recsig);
+    savsig[SIGSEGV] = signal(SIGSEGV, recsig);
+    savsig[SIGTERM] = signal(SIGTERM, recsig);
+# endif
+
+// set SIGWINCH handler to default so our workflow is not
+// triggered on user resizing of window
+# if defined(SIGWINCH) && defined(SIG_DFL)
+    signal(SIGWINCH, SIG_DFL);
+# endif
+}
+
+static void popsig(void) {
+# if !defined(_WIN32)
+    int i;
+    for (i = 1; i < NUM_SIG; i++) {
+        if (i == SIGUSR1 || i == SIGUSR2 || i == SIGKILL) {
+          continue;
+        }
+        sigaction(i, &savsig[i], NULL);
+    }
+# else
+    signal(SIGABRT, savsig[SIGABRT]);
+    signal(SIGFPE, savsig[SIGFPE]);
+    signal(SIGILL, savsig[SIGILL]);
+    signal(SIGINT, savsig[SIGINT]);
+    signal(SIGSEGV, savsig[SIGSEGV]);
+    signal(SIGTERM, savsig[SIGTERM]);
+# endif
+}
+
+static void recsig(int signal) {
+    intr_signal = signal;
+}
+
+/* Console management functions */
+int openssl_console_open(void) {
+    CRYPTO_STATIC_MUTEX_lock_write(&console_global_mutex);
+    is_a_tty = 1;
+
+    # if !defined(_WIN32)
+    if ((tty_in = fopen(DEV_TTY, "r")) == NULL) {
+        tty_in = stdin;
+    }
+    if ((tty_out = fopen(DEV_TTY, "w")) == NULL) {
+        tty_out = stderr;
+    }
+    if (TTY_get(fileno(tty_in), &tty_orig) == -1) {
+        if (errno == ENOTTY || errno == EINVAL || errno == ENXIO || errno == EIO
+          || errno == EPERM || errno == ENODEV) {
+            is_a_tty = 0;
+          } else {
+              OPENSSL_PUT_ERROR(PEM, ERR_R_INTERNAL_ERROR);
+              return 0;
```

Comment:
If you return an error, then you must release the lock as it appears that the caller is unaware that a write lock was acquired. (See comment above about moving the locking up to the caller.)